### DISCUSSION TITLE
Best way to read files during Benchmark run

### DISCUSSION THREAD

**Discussion by lbussell** [dotnet/BenchmarkDotNet#2472] [Q&A] [answered]
Hi, I would like to read text input from a file in my Benchmark runs, but I'm having trouble with using relative paths. My code reads the files just fine outside of the Benchmark, but it can't find the files when the Benchmark runs. I get a DirectoryNotFoundException:

```
Could not find a part of the path 'C:\s\advent-of-code\src\AdventOfCode.Cli\bin\Release\net8.0\36ff3b4a-91a3-444c-bb7e-fc473289d2fe\bin\Release\net8.0\Inputs\00.txt'
```

The way I currently have it set up is referring to the file I want to read relative to the `.csproj` I'm running from. I understand that BenchmarkDotNet generates its own project files, so I'm wondering how I can pass the files down to the BenchmarkDotNet project correctly so they're available when its running. My project has no issue running with `dotnet run` or `dotnet run -c Release`.

Here is my benchmark class: https://github.com/lbussell/advent-of-code/blob/cc638c0648739df800c8045053f9f058e9fcc377/src/AdventOfCode.Cli/SolutionBenchmark.cs#L25

And here is where I read the file. It is read as part of the constructor for each class implementation that I want to run a benchmark on: https://github.com/lbussell/advent-of-code/blob/cc638c0648739df800c8045053f9f058e9fcc377/src/AdventOfCode.Cli/Solutions/SolutionWithTextInput.cs#L7

I would appreciate any advice on getting this to work.

**Comment by mawosoft** [comment, marked as answer]
You can use a local helper function like:
```cs
static string WhereAmI([CallerFilePath] string callerFilePath = "") => callerFilePath;
```
and calculate the path relative to the source code file that's calling the function, e.g.
```cs
string basePath = Path.GetDirectoryName(WhereAmI());
string targetPath = Path.Combine(basePath, "relative/path/to/file.txt");
```

For a more complex approach, dealing with deterministic paths, CI, etc. see
https://github.com/mawosoft/MissingCoverage/blob/master/tests/testdata/src/TestDataDirectory.cs

**Reply by lbussell** [reply]
This is awesome, it works great. Here's the commit that I made in case anyone runs into the same issue and wants to see another example: https://github.com/lbussell/advent-of-code/commit/717a1e697ca445a1f6550d2db6b70960b84d02f2
